File: AC/library/ilib/aczip/aczip.py
import gzip
import tarfile
import os
import io
from pathlib import Path
from concurrent.futures import ThreadPoolExecutor

class ACZip:
    """4-bit protocol archiver for AC projects"""

    @staticmethod
    def compress(path, parallel=True):
        """Compress directory using 4-bit tagging protocol"""
        files = []
        for root, dirs, filenames in os.walk(path):
            for filename in filenames:
                filepath = os.path.join(root, filename)
                with open(filepath, 'rb') as f:
                    data = f.read()
                    rel_path = os.path.relpath(filepath, path)
                    files.append((rel_path, data))

        # Serialize with 4-bit tags
        serialized = b'ACZP'  # Magic header
        for idx, (filepath, data) in enumerate(files):
            tag = ACZip._generate_tag(idx)
            serialized += tag.encode() + f'{len(data)}:'.encode() + data

        # Gzip compress
        return gzip.compress(serialized, compresslevel=6)

    @staticmethod
    def decompress(data, output_path):
        """Decompress ACZip archive"""
        decompressed = gzip.decompress(data)
        serialized = decompressed.decode('latin-1', errors='ignore')

        if not serialized.startswith('ACZP'):
            raise ValueError("Invalid ACZip file")

        pos = 4
        while pos < len(serialized):
            colon = serialized.find(':', pos + 4)
            tag = serialized[pos:pos+4]
            size = int(serialized[pos+4:colon])
            pos = colon + 1

            file_data = serialized[pos:pos+size].encode('latin-1')
            pos += size

            # Reconstruct path from tag
            file_path = os.path.join(output_path, tag)
            os.makedirs(os.path.dirname(file_path), exist_ok=True)

            with open(file_path, 'wb') as f:
                f.write(file_data)

    @staticmethod
    def compress_hdd(path):
        """Compress optimized for HDD (sequential)"""
        return ACZip.compress(path, parallel=False)

    @staticmethod
    def compress_sata(path):
        """Compress optimized for SATA (balanced)"""
        return ACZip.compress(path, parallel=True)

    @staticmethod
    def get_ratio(original_size, compressed_size):
        """Calculate compression ratio"""
        if original_size == 0:
            return 0.0
        return (compressed_size * 100.0) / original_size

    @staticmethod
    def _generate_tag(index):
        """Generate 4-bit tag for file index"""
        if index < 15:
            return f"0x{index:X}"
        folder = (index // 15) + 1
        file_idx = (index % 15) + 1
        return f"1x{folder:02d}.0x{file_idx:02X}"


class ACTar:
    """TAR archive creation and extraction"""

    @staticmethod
    def create(path):
        """Create TAR archive from path"""
        tar_buffer = io.BytesIO()
        with tarfile.open(fileobj=tar_buffer, mode='w') as tar:
            tar.add(path, arcname=os.path.basename(path))
        return tar_buffer.getvalue()

    @staticmethod
    def extract(data, output_path):
        """Extract TAR archive"""
        tar_buffer = io.BytesIO(data)
        with tarfile.open(fileobj=tar_buffer, mode='r') as tar:
            tar.extractall(path=output_path)


class ACGzip:
    """GZIP compression and decompression"""

    @staticmethod
    def compress(data, level=6):
        """Compress data with gzip"""
        return gzip.compress(data, compresslevel=level)

    @staticmethod
    def decompress(data):
        """Decompress gzip data"""
        return gzip.decompress(data)

File: AC/library/ilib/aczip/aczip.js
const fs = require('fs');
const path = require('path');
const zlib = require('zlib');
const tar = require('tar');

class ACZip {
    /**
     * Compress directory: each file compressed individually in parallel
     */
    static compress(dirPath, parallel = true) {
        const files = [];

        // Collect all files
        const walk = (dir) => {
            const items = fs.readdirSync(dir);
            for (const item of items) {
                const fullPath = path.join(dir, item);
                const stat = fs.statSync(fullPath);
                if (stat.isFile()) {
                    const data = fs.readFileSync(fullPath);
                    const relPath = path.relative(dirPath, fullPath);
                    files.push({ path: relPath, data });
                } else if (stat.isDirectory()) {
                    walk(fullPath);
                }
            }
        };
        walk(dirPath);

        // Compress each file individually
        return new Promise((resolve, reject) => {
            const compressedFiles = [];
            let completed = 0;

            if (parallel && files.length > 1) {
                // Parallel compression
                files.forEach((file, idx) => {
                    zlib.gzip(file.data, { level: 6 }, (err, compressed) => {
                        if (err) return reject(err);

                        compressedFiles[idx] = {
                            tag: ACZip._generateTag(idx),
                            compressed
                        };

                        if (++completed === files.length) {
                            resolve(ACZip._buildArchive(compressedFiles));
                        }
                    });
                });
            } else {
                // Sequential compression
                const compress = (idx) => {
                    if (idx >= files.length) {
                        return resolve(ACZip._buildArchive(compressedFiles));
                    }

                    zlib.gzip(files[idx].data, { level: 6 }, (err, compressed) => {
                        if (err) return reject(err);

                        compressedFiles.push({
                            tag: ACZip._generateTag(idx),
                            compressed
                        });

                        compress(idx + 1);
                    });
                };
                compress(0);
            }
        });
    }

    /**
     * Decompress ACZip v2 archive
     */
    static decompress(data, outputPath) {
        return new Promise((resolve, reject) => {
            // Check magic header
            if (data.length < 8 || data.toString('utf8', 0, 4) !== 'ACZ2') {
                return reject(new Error('Invalid ACZip file'));
            }

            let pos = 4;
            const fileCount = data.readUInt32LE(pos);
            pos += 4;

            fs.mkdir(outputPath, { recursive: true }, (err) => {
                if (err) return reject(err);

                const promises = [];

                for (let i = 0; i < fileCount; i++) {
                    // Tag
                    const tag = data.toString('utf8', pos, pos + 4);
                    pos += 4;

                    // Compressed size
                    const compSize = data.readUInt32LE(pos);
                    pos += 4;

                    // Compressed data
                    const compressed = data.slice(pos, pos + compSize);
                    pos += compSize;

                    // Decompress
                    promises.push(
                        new Promise((res, rej) => {
                            zlib.gunzip(compressed, (err, decompressed) => {
                                if (err) return rej(err);

                                const filePath = path.join(outputPath, tag);
                                fs.mkdir(path.dirname(filePath), { recursive: true }, (err) => {
                                    if (err) return rej(err);
                                    fs.writeFile(filePath, decompressed, (err) => {
                                        if (err) rej(err);
                                        else res();
                                    });
                                });
                            });
                        })
                    );
                }

                Promise.all(promises).then(resolve).catch(reject);
            });
        });
    }

    /**
     * Compress optimized for HDD
     */
    static compressHDD(dirPath) {
        return ACZip.compress(dirPath, false);
    }

    /**
     * Compress optimized for SATA
     */
    static compressSATA(dirPath) {
        return ACZip.compress(dirPath, true);
    }

    /**
     * Calculate compression ratio
     */
    static getRatio(originalSize, compressedSize) {
        if (originalSize === 0) return 0;
        return (compressedSize * 100.0) / originalSize;
    }

    /**
     * Generate 4-byte tag for file index
     */
    static _generateTag(index) {
        if (index < 15) {
            return `0x${index.toString(16).toUpperCase()}`.padEnd(4, ' ');
        }
        const folder = Math.floor(index / 15) + 1;
        const fileIdx = (index % 15) + 1;
        return `1x${folder.toString().padStart(2, '0')}.0x${fileIdx.toString(16).toUpperCase().padStart(2, '0')}`.slice(0, 4);
    }

    /**
     * Build v2 archive from compressed files
     */
    static _buildArchive(compressedFiles) {
        let result = Buffer.from('ACZ2');
        result = Buffer.concat([result, Buffer.alloc(4)]);
        result.writeUInt32LE(compressedFiles.length, 4);

        for (const file of compressedFiles) {
            const tag = Buffer.from(file.tag, 'utf8');
            const sizeBuffer = Buffer.alloc(4);
            sizeBuffer.writeUInt32LE(file.compressed.length);

            result = Buffer.concat([result, tag, sizeBuffer, file.compressed]);
        }

        return result;
    }
}

class ACTar {
    /**
     * Create TAR archive
     */
    static create(dirPath) {
        return new Promise((resolve, reject) => {
            const chunks = [];
            const stream = tar.create(
                { gzip: false },
                [dirPath]
            );

            stream.on('data', (chunk) => chunks.push(chunk));
            stream.on('end', () => resolve(Buffer.concat(chunks)));
            stream.on('error', reject);
        });
    }

    /**
     * Extract TAR archive
     */
    static extract(data, outputPath) {
        return new Promise((resolve, reject) => {
            const stream = tar.extract({ cwd: outputPath });
            stream.on('finish', resolve);
            stream.on('error', reject);
            stream.end(data);
        });
    }
}

class ACGzip {
    /**
     * Compress with gzip
     */
    static compress(data, level = 6) {
        return new Promise((resolve, reject) => {
            zlib.gzip(data, { level }, (err, compressed) => {
                if (err) reject(err);
                else resolve(compressed);
            });
        });
    }

    /**
     * Decompress gzip data
     */
    static decompress(data) {
        return new Promise((resolve, reject) => {
            zlib.gunzip(data, (err, decompressed) => {
                if (err) reject(err);
                else resolve(decompressed);
            });
        });
    }
}

module.exports = { ACZip, ACTar, ACGzip };
